be/src/storage/tablet.cpp

```diff
@@ -1170,4 +1198,122 @@ Version Tablet::max_version() const {
     }
 }
 
+void Tablet::update_tablet_compaction_context() {
+    if (_updates || _state == TABLET_NOTREADY || !is_used() || !init_succeeded()) {
+        _compaction_context.reset();
+        return;
+    }
+    if (!_check_versions_completeness()) {
+        _compaction_context.reset();
+        // when versions are not complete, just return
+        return;
+    }
+    std::unique_ptr<CompactionContext> compaction_context = _get_compaction_context();
+    std::unique_ptr<CompactionPolicy> compaction_policy =
+            CompactionUtils::create_compaction_policy(compaction_context.get());
+    if (compaction_policy->need_compaction()) {
+        _compaction_context.swap(compaction_context);
+    } else {
+        _compaction_context.reset();
+    }
+}
+
+bool Tablet::_is_compacted_singleton(Rowset* rowset) {
+    DCHECK(rowset) << "invalid rowset";
+    return rowset->num_segments() > 1 && rowset->rowset_meta()->segments_overlap() == NONOVERLAPPING;
+}
+
+// protected by meta lock
+std::unique_ptr<CompactionContext> Tablet::_get_compaction_context() {
```

Comment:
Where is the meta lock?

Reply:
outside this function, In the calling function. Now it is in get_compaction_candidates